src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1925,9 +1940,21 @@ SingleShardModifyTaskList(Query *query, uint64 jobId, List *relationShardList,
 							   "and modify a reference table")));
 	}
 
-	pg_get_query_def(query, queryString);
+	/*
+	 * For performance reasons, we skip generating the queryString for local
+	 * fast path queries. With this, we can avoid deparsing during the execution
+	 * as well.
+	 */
+	if (localFastPathQuery)
+	{
+		queryString = NULL;
+	}
+	else
+	{
+		pg_get_query_def(query, queryString);
```

Comment:
High level question: Is there any benefit we get from generating the query string here as opposed to in the executor?

Reply:
I added a commit that tries to do this by storing a copy of the query in the Task, instead of a string. Overal it seems to work pretty fine. The only thing that breaks badly seems are prepared statements without parameters. I expect it's related to the string memory being in the wrong memory context.